ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,902 @@
+    }
+
+    private string getActionString() {
+      result = method.getKeywordArgument("to").(StringlikeLiteral).getValueText()
+      or
+      method.getKeywordArgument("to").(MethodCall).getMethodName() = "redirect" and
+      result = "<redirect>#<redirect>"
+    }
+
+    override string getAction() {
+      // get "/photos", action: "index"
+      result = method.getKeywordArgument("action").getValueText()
+      or
+      not exists(method.getKeywordArgument("action")) and
+      (
+        // get "/photos", to: "photos#index"
+        // get "/photos", to: redirect("some_url")
+        result = extractAction(this.getActionString())
+        or
+        // resources :photos, only: [] do
+        //   get "/", to: "index"
+        // end
+        not exists(extractAction(this.getActionString())) and result = this.getActionString()
+        or
+        // get :some_action
+        not exists(this.getActionString()) and
+        result = method.getArgument(0).(StringlikeLiteral).getValueText()
+      )
+    }
+
+    override string getHTTPMethod() { result = method.getMethodName().toString() }
+  }
+
+  /**
+   * A route generated by a call to `resources`.
+   *
+   * ```ruby
+   * resources :photos
+   * ```
+   * This creates eight routes, equivalent to the following code:
+   * ```ruby
+   * get "/photos", to: "photos#index"
+   * get "/photos/new", to: "photos#new"
+   * post "/photos", to: "photos#create"
+   * get "/photos/:id", to: "photos#show"
+   * get "/photos/:id/edit", to: "photos#edit"
+   * patch "/photos/:id", to: "photos#update"
+   * put "/photos/:id", to: "photos#update"
+   * delete "/photos/:id", to: "photos#delete"
+   * ```
+   *
+   * `resources` can take a block. Any routes defined inside the block will inherit a path component of
+   * `/<resource>/:<resource>_id`. For example:
+   *
+   * ```ruby
+   * resources :photos do
+   *   get "/foo", to: "photos#foo"
+   * end
+   * ```
+   * This creates the eight default routes, plus one more, which is nested under "/photos/:photo_id", equivalent to:
+   * ```ruby
+   * get "/photos/:photo_id/foo", to: "photos#foo"
+   * ```
+   */
+  class ResourcesRoute extends Route, TResourcesRoute {
+    RouteBlock parent;
+    string resource;
+    string action;
+    string httpMethod;
+    string pathComponent;
+
+    ResourcesRoute() {
+      this = TResourcesRoute(parent, method, action) and
+      resource = method.getArgument(0).(StringlikeLiteral).getValueText() and
+      isDefaultResourceRoute(resource, httpMethod, pathComponent, action)
+    }
+
+    override string getAPrimaryQlClass() { result = "ResourcesRoute" }
+
+    override RouteBlock getParentBlock() { result = parent }
+
+    override string getLastPathComponent() { result = pathComponent }
+
+    override string getLastControllerComponent() { result = method.getArgument(0).getValueText() }
+
+    override string getAction() { result = action }
+
+    override string getHTTPMethod() { result = httpMethod }
```

Comment:
I noticed this returning `patch` for a call like this:

```rb
resources :foo,
    path: "/foo/:foo_id/bar",
    only: %w(new create edit update destroy),
    ...
```

Reply:
[`PATCH`](https://datatracker.ietf.org/doc/html/rfc2068#section-19.6.1.1) is a valid, if rarely used, HTTP method which acts similarly to `PUT`, but is intended for partial updates to a resource (whereas `PUT` is for completely replacing the resource). Rails prefers `PATCH` for updates because it is a better fit semantically. (https://guides.rubyonrails.org/v6.0/4_0_release_notes.html)